server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +347,278 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    /**
+     * iFace which implements the query timeout / cancellation logic
+     */
+    public interface QueryCancellable {
+
+        void checkCancelled();
+    }
+
+    /**
+     * Wraps an {@link IndexReader} with a {@link QueryCancellable}.
+     */
+    static class CancellableDirectoryReader extends FilterDirectoryReader {
```

Comment:
nit: This class deserves its own file. Is there a specific reason to rename from `ExitableDirectoryReader` ?

Reply:
Only this class or also include the underlying wrapper classes?

Reply:
Also if we move it to its own file then it needs to know of the `MutableQueryTimeout`.

Reply:
> Only this class or also include the underlying wrapper classes?

All of it, yes.

> Also if we move it to its own file then it needs to know of the MutableQueryTimeout.

Why that ? The idea is to have a simple interface in the reader and to use the mutable impl only for the searcher. You can create one in the ctr like this:
````
 public ContextIndexSearcher(IndexReader reader,
                                Similarity similarity,
                                QueryCache queryCache,
                                QueryCachingPolicy queryCachingPolicy) throws IOException {
        this(reader, similarity, queryCache, queryCachingPolicy, new MutableQueryTimeout());
    }

    private ContextIndexSearcher(IndexReader reader,
                                Similarity similarity,
                                QueryCache queryCache,
                                QueryCachingPolicy queryCachingPolicy,
                                MutableQueryTimeout timeout) throws IOException {
        super(ExitableDirectoryReader.wrap((DirectoryReader) reader, timeout));
        this.queryTimeout = timeout;
        setSimilarity(similarity);
        setQueryCache(queryCache);
        setQueryCachingPolicy(queryCachingPolicy);
    }
````
?